src/coreclr/jit/valuenum.cpp

```diff
@@ -8743,6 +8834,13 @@ void Compiler::fgValueNumberTree(GenTree* tree)
                     // Note VNF_PtrToStatic statics are currently always "simple".
                     fgValueNumberFieldLoad(tree, /* baseAddr */ nullptr, fldSeq, offset);
                 }
+                else if (tree->OperIs(GT_IND) && fgValueNumberConstStringElemLoad(tree->AsIndir()))
+                {
+                    tree->gtFlags |= GTF_IND_NONFAULTING | GTF_IND_INVARIANT;
+                    tree->gtFlags &= ~GTF_EXCEPT;
+                    // Update side-effect flags of parent trees if needed
+                    gtUpdateStmtSideEffects(stmt);
```

Comment:
It is not right to update side effects in VN like that. The problem this solves should be solved in some other way.

Reply:
@SingleAccretion Any recommendation and what's wrong with it? Unfortunately, I don't see a better option and this is required for this PR.
Alternative option I see is to define a new VN with "known-non-null" property and then fold it later e.g. at assertprop.

Reply:
> Any recommendation

From the discussion yesterday - did making VN-based constant propagation run in post-order not work?

> what's wrong with it

It's not in VN's contract to modify IR like this. Certainly, if you allow this, all sorts of things become possible (for example, here we would just bash the node to a `CNS_INT` directly).

Reply:
Thanks! Reverted the change, the suggestion about post-order works but I decided to postpone it to a separate PR so I can see if it worth the effort around TP and diffs (and planned a few experiments around that)